README
This package contains a base64 encoder/decoder and a quoted-printable
encoder/decoder.  These encoding methods are specified in RFC 2045 -
MIME (Multipurpose Internet Mail Extensions).

The base64 encoding is designed to represent arbitrary sequences of
octets in a form that need not be humanly readable. A 65-character
subset ([A-Za-z0-9+/=]) of US-ASCII is used, enabling 6 bits to be
represented per printable character.

The quoted-printable encoding is intended to represent data that
largely consists of bytes that correspond to printable characters in
the ASCII character set.  Non-printable characters (as defined by
enghlish americans) are represented by a triplet consisting of the
character "=" followed by two hexadecimal digits.

The MIME::Base64 and MIME::QuotedPrint modules used to be part of
libwww-perl package.  They are now distributed separately (this
package).  The main news is that the base64 encoder is implemented by
XS functions.  This makes it about 20 times faster than the old
implementation in perl.

In order to install and use this package you will need Perl version
5.002 or better.

   perl Makefile.PL
   make
   make test
   make install

Copyright 1995-1997 Gisle Aas.

This library is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.


Changes
2010-10-26   Gisle Aas <gisle@ActiveState.com>

   Release 3.13

   The fix in v3.12 to try to preserve the SvUTF8 flag was buggy
   and actually managed to set the flag on strings that did not
   have it originally.



2010-10-25   Gisle Aas <gisle@ActiveState.com>

   Release 3.12

   Don't change SvUTF8 flag on the strings encoded [RT#60105]

   Documentation tweaks



2010-10-24   Gisle Aas <gisle@ActiveState.com>

   Release 3.11

   Provide encode_base64url and decode_base64url functions to process
   the base64 scheme for "URL applications".

   The decode_base64() does not issue warnings on suspect input data
   any more.



2010-10-11   Gisle Aas <gisle@ActiveState.com>

   Release 3.10

   Provide functions to calculate the length of encoded and decoded
   base64 strings [RT#62404]



2010-01-25   Gisle Aas <gisle@ActiveState.com>

   Release 3.09

   The Quoted-Printable encoder would sometimes output lines
   that were 77 characters long.  The max line length should be 76.
   [RT#53919]



2009-06-09   Gisle Aas <gisle@ActiveState.com>

   Release 3.08

   Jarkko Hietaniemi (1):
      EBCDIC changes from core

   Nicholas Clark (1):
      Get rid of the PERL_CORE hacks



2005-11-30   Gisle Aas <gisle@ActiveState.com>

   Release 3.07

   Use a Makefile.PL that is also suitable for core perl.



2005-11-26   Gisle Aas <gisle@ActiveState.com>

   Release 3.06

   Documentation tweaks.

   use XSLoader; perl-5.6 now required.

   Some consting from bleadperl.

   Unbundled the {en,de}code-{base64,qp} utility scripts.
   These are now found in the MIME-Base64-Scripts package.



2004-09-20   Gisle Aas <gisle@ActiveState.com>

   Release 3.05

   Steve Hay <steve.hay@uk.radan.com> found the warn test broken
   on Windows and provided a fix.



2004-09-18   Gisle Aas <gisle@ActiveState.com>

   Release 3.04
   
   Fixed the bad-sv.t test script to actually contain the
   correct expected result as of v3.02.



2004-08-25   Gisle Aas <gisle@ActiveState.com>

   Release 3.03

   Forgot to increment version number in MIME::QuotedPrint even
   if its interface changed in 3.02.  As a result you will now
   need to require MIME::QuotedPrint 3.03 if you want to ensure
   it provides the binmode interface.



2004-08-24   Gisle Aas <gisle@ActiveState.com>

   Release 3.02

   The encode_qp() function now takes an optional third argument
   to select binary encoding mode.
   <https://rt.cpan.org/Ticket/Display.html?id=7456>

   The result of encode_qp($non_empty, $eol) will now always be
   $eol terminated.  If the string to encode does not end with "\n"
   then a soft line break is appended to the result.  As an example
   encode_qp("foo") used to be encoded as "foo", but now encodes as
   "foo=\n".



2004-03-29   Gisle Aas <gisle@ActiveState.com>

   Release 3.01

   By compiling the extension with PERL_NO_GET_CONTEXT we can
   make it slightly faster on a threaded perl.  No change on a
   regular perl.  Patch provided by Beau E. Cox <beau@beaucox.com>.

   Fixed missing ";" with assert.   Patch provided by
   Brendan O'Dea <bod@debian.org>.



2004-01-14   Gisle Aas <gisle@ActiveState.com>

   Release 3.00

   Drop the pure Perl implementations of the encoders and
   decoders.  They are bloat that hides real problems in
   the XS implementations.  I will re-release them separately
   in the new MIME-Base64-Perl distribution.

   The 'gcc -Wall' fix in 2.22 broke support for perl5.005,
   as the isXDIGIT() macro is not available in that perl.
   This problem has now been fixed.



2004-01-08   Gisle Aas <gisle@ActiveState.com>

   Release 2.23

   Documentation fixes by Paul Croome <Paul.Croome@softwareag.com>.



2004-01-08   Gisle Aas <gisle@ActiveState.com>

   Release 2.22

   Fix 'gcc -Wall' complaints.



2003-10-09   Gisle Aas <gisle@ActiveState.com>

   Release 2.21

   Documentation tweaks.

   Don't rely on SvEND(sv) == '\0' as discussed in the perl5-porters
   mailing list thread that starts with
   http://www.xray.mpe.mpg.de/mailing-lists/perl5-porters/2003-10/msg00258.html

   Should now pass test suite even without XS support.

   Perl v5.005 or better is now required.



2003-05-13   Gisle Aas <gisle@ActiveState.com>

   Release 2.20

   decode_qp() recognize soft whitespace when there is whitespace
   between the '=' and the '\n'.



2003-05-13   Gisle Aas <gisle@ActiveState.com>

   Release 2.19

   decode_qp() did eat up all trailing whitespace in the string decoded.
   Only whitespace in front of "\n" should go.

   Win32 fix for t/warn.t by Reini Urban <rurban@x-ray.at>.



2003-03-09   Gisle Aas <gisle@ActiveState.com>

   Release 2.18

   Fix up INSTALLDIRS for perl-5.8 and newer.



2003-03-09   Gisle Aas <gisle@ActiveState.com>

   Release 2.17

   Make it reliable to disable base64 decoding warnings by
   resetting $^W in recent perls.  Would really like to be
   able to do real lexical warnings but the current mechanism
   does not seems suitable for XS code.

   Passing "" as $eol to encode_qp() disable soft line
   breaks as well.

   Sync up with changes in bleadperl:
       - safer patchlevel.h include
       - bad cast



2003-01-05   Gisle Aas <gisle@ActiveState.com>

   Release 2.16

   Fixed the encode_qp() line breaking code.  It sometimes
   made lines longer than 76 chars and it could even get into
   an infinite loop on certain inputs.



2003-01-03   Gisle Aas <gisle@ActiveState.com>

   Release 2.15

   Fixed the XS based decode_qp() for strings where a =XX
   sequence was followed by digits.

   Faster encode_qp() for long strings with lots of chars
   that need escaping.

   The old_decode_base64() function introduced in 2.13
   was returning undef for empty input on olders perls.
   This problem has been fixed.



2003-01-01   Gisle Aas <gisle@ActiveState.com>

   Release 2.14

   MIME::QuotedPrint functions now also implemented using XS
   which make them faster.  2-3 times faster when encoding line by

QuotedPrint.pm
package MIME::QuotedPrint;

use strict;
use vars qw(@ISA @EXPORT $VERSION);

require Exporter;
@ISA = qw(Exporter);
@EXPORT = qw(encode_qp decode_qp);

$VERSION = "3.11";

use MIME::Base64;  # will load XS version of {en,de}code_qp()

*encode = \&encode_qp;
*decode = \&decode_qp;

1;

__END__

=head1 NAME

MIME::QuotedPrint - Encoding and decoding of quoted-printable strings

=head1 SYNOPSIS

 use MIME::QuotedPrint;

 $encoded = encode_qp($decoded);
 $decoded = decode_qp($encoded);

=head1 DESCRIPTION

This module provides functions to encode and decode strings into and from the
quoted-printable encoding specified in RFC 2045 - I<MIME (Multipurpose
Internet Mail Extensions)>.  The quoted-printable encoding is intended
to represent data that largely consists of bytes that correspond to
printable characters in the ASCII character set.  Each non-printable
character (as defined by English Americans) is represented by a
triplet consisting of the character "=" followed by two hexadecimal
digits.

The following functions are provided:

=over 4

=item encode_qp( $str)

=item encode_qp( $str, $eol)

=item encode_qp( $str, $eol, $binmode )

This function returns an encoded version of the string ($str) given as
argument.

The second argument ($eol) is the line-ending sequence to use.  It is
optional and defaults to "\n".  Every occurrence of "\n" is replaced
with this string, and it is also used for additional "soft line
breaks" to ensure that no line end up longer than 76 characters.  Pass
it as "\015\012" to produce data suitable for external consumption.
The string "\r\n" produces the same result on many platforms, but not
all.

The third argument ($binmode) will select binary mode if passed as a
TRUE value.  In binary mode "\n" will be encoded in the same way as
any other non-printable character.  This ensures that a decoder will
end up with exactly the same string whatever line ending sequence it
uses.  In general it is preferable to use the base64 encoding for
binary data; see L<MIME::Base64>.

An $eol of "" (the empty string) is special.  In this case, no "soft
line breaks" are introduced and binary mode is effectively enabled so
that any "\n" in the original data is encoded as well.

=item decode_qp( $str )

This function returns the plain text version of the string given
as argument.  The lines of the result are "\n" terminated, even if
the $str argument contains "\r\n" terminated lines.

=back


If you prefer not to import these routines into your namespace, you can
call them as:

  use MIME::QuotedPrint ();
  $encoded = MIME::QuotedPrint::encode($decoded);
  $decoded = MIME::QuotedPrint::decode($encoded);

Perl v5.8 and better allow extended Unicode characters in strings.
Such strings cannot be encoded directly, as the quoted-printable
encoding is only defined for single-byte characters.  The solution is
to use the Encode module to select the byte encoding you want.  For
example:

    use MIME::QuotedPrint qw(encode_qp);
    use Encode qw(encode);

    $encoded = encode_qp(encode("UTF-8", "\x{FFFF}\n"));
    print $encoded;

=head1 COPYRIGHT

Copyright 1995-1997,2002-2004 Gisle Aas.

This library is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=head1 SEE ALSO

L<MIME::Base64>

=cut


Base64.pm
package MIME::Base64;

use strict;
use vars qw(@ISA @EXPORT @EXPORT_OK $VERSION);

require Exporter;
@ISA = qw(Exporter);
@EXPORT = qw(encode_base64 decode_base64);
@EXPORT_OK = qw(encode_base64url decode_base64url encoded_base64_length decoded_base64_length);

$VERSION = '3.11';

require XSLoader;
XSLoader::load('MIME::Base64', $VERSION);

*encode = \&encode_base64;
*decode = \&decode_base64;

sub encode_base64url {
    my $e = encode_base64(shift, "");
    $e =~ s/=+\z//;
    $e =~ tr[+/][-_];
    return $e;
}

sub decode_base64url {
    my $s = shift;
    $s =~ tr[-_][+/];
    $s .= '=' while length($s) % 4;
    return decode_base64($s);
}

1;

__END__

=head1 NAME

MIME::Base64 - Encoding and decoding of base64 strings

=head1 SYNOPSIS

 use MIME::Base64;

 $encoded = encode_base64('Aladdin:open sesame');
 $decoded = decode_base64($encoded);

=head1 DESCRIPTION

This module provides functions to encode and decode strings into and from the
base64 encoding specified in RFC 2045 - I<MIME (Multipurpose Internet
Mail Extensions)>. The base64 encoding is designed to represent
arbitrary sequences of octets in a form that need not be humanly
readable. A 65-character subset ([A-Za-z0-9+/=]) of US-ASCII is used,
enabling 6 bits to be represented per printable character.

The following primary functions are provided:

=over 4

=item encode_base64( $bytes )

=item encode_base64( $bytes, $eol );

Encode data by calling the encode_base64() function.  The first
argument is the byte string to encode.  The second argument is the
line-ending sequence to use.  It is optional and defaults to "\n".  The
returned encoded string is broken into lines of no more than 76
characters each and it will end with $eol unless it is empty.  Pass an
empty string as second argument if you do not want the encoded string
to be broken into lines.

The function will croak with "Wide character in subroutine entry" if $bytes
contains characters with code above 255.  The base64 encoding is only defined
for single-byte characters.  Use the Encode module to select the byte encoding
you want.

=item decode_base64( $str )

Decode a base64 string by calling the decode_base64() function.  This
function takes a single argument which is the string to decode and
returns the decoded data.

Any character not part of the 65-character base64 subset is
silently ignored.  Characters occurring after a '=' padding character
are never decoded.

=back

If you prefer not to import these routines into your namespace, you can
call them as:

    use MIME::Base64 ();
    $encoded = MIME::Base64::encode($decoded);
    $decoded = MIME::Base64::decode($encoded);

Additional functions not exported by default:

=over 4

=item encode_base64url( $bytes )

=item decode_base64url( $str )

Encode and decode according to the base64 scheme for "URL applications" [1].
This is a variant of the base64 encoding which does not use padding, does not
break the string into multiple lines and use the characters "-" and "_" instead
of "+" and "/" to avoid using reserved URL characters.

=item encoded_base64_length( $bytes )

=item encoded_base64_length( $bytes, $eol )

Returns the length that the encoded string would have without actually
encoding it.  This will return the same value as C<< length(encode_base64($bytes)) >>,
but should be more efficient.

=item decoded_base64_length( $str )

Returns the length that the decoded string would have without actually
decoding it.  This will return the same value as C<< length(decode_base64($str)) >>,
but should be more efficient.

=back

=head1 EXAMPLES

If you want to encode a large file, you should encode it in chunks
that are a multiple of 57 bytes.  This ensures that the base64 lines
line up and that you do not end up with padding in the middle. 57
bytes of data fills one complete base64 line (76 == 57*4/3):

   use MIME::Base64 qw(encode_base64);

   open(FILE, "/var/log/wtmp") or die "$!";
   while (read(FILE, $buf, 60*57)) {
       print encode_base64($buf);
   }

or if you know you have enough memory

   use MIME::Base64 qw(encode_base64);
   local($/) = undef;  # slurp
   print encode_base64(<STDIN>);

The same approach as a command line:

   perl -MMIME::Base64 -0777 -ne 'print encode_base64($_)' <file

Decoding does not need slurp mode if every line contains a multiple
of four base64 chars:

   perl -MMIME::Base64 -ne 'print decode_base64($_)' <file

Perl v5.8 and better allow extended Unicode characters in strings.
Such strings cannot be encoded directly, as the base64
encoding is only defined for single-byte characters.  The solution is
to use the Encode module to select the byte encoding you want.  For
example:

    use MIME::Base64 qw(encode_base64);
    use Encode qw(encode);

    $encoded = encode_base64(encode("UTF-8", "\x{FFFF}\n"));
    print $encoded;

=head1 COPYRIGHT

Copyright 1995-1999, 2001-2004, 2010 Gisle Aas.

This library is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

Distantly based on LWP::Base64 written by Martijn Koster
<m.koster@nexor.co.uk> and Joerg Reichelt <j.reichelt@nexor.co.uk> and
code posted to comp.lang.perl <3pd2lp$6gf@wsinti07.win.tue.nl> by Hans
Mulder <hansm@wsinti07.win.tue.nl>

The XS implementation uses code from metamail.  Copyright 1991 Bell
Communications Research, Inc. (Bellcore)

=head1 SEE ALSO

L<MIME::QuotedPrint>

[1] L<http://en.wikipedia.org/wiki/Base64#URL_applications>

=cut
